src/sys/statfs.rs

```diff
@@ -1,22 +1,432 @@
+        not(any(
+            target_os = "ios",
+            target_os = "macos",
+            target_os = "android",
+            target_os = "freebsd"
+        ))
+    ))]
+    pub fn blocks_available(&self) -> libc::uint32_t {
+        self.0.f_bavail
+    }
+
+    #[cfg(target_os = "freebsd")]
+    pub fn blocks_available(&self) -> libc::int64_t {
+        self.0.f_bavail
+    }
+
+    /// Total file nodes in filesystem
+    #[cfg(any(
+        target_os = "ios",
+        target_os = "macos",
+        target_os = "android",
+        target_os = "freebsd",
+        target_env = "musl",
+        target_pointer_width = "64"
+    ))]
+    pub fn files(&self) -> libc::uint64_t {
+        self.0.f_files
+    }
+
+    /// Total file nodes in filesystem
+    #[cfg(all(
+        target_pointer_width = "32",
+        not(any(
+            target_os = "ios",
+            target_os = "macos",
+            target_os = "android",
+            target_os = "freebsd",
+            target_env = "musl"
+        ))
+    ))]
+    pub fn files(&self) -> libc::uint32_t {
+        self.0.f_files
+    }
+
+    /// Free file nodes in filesystem
+    #[cfg(all(
+        any(
+            target_os = "ios",
+            target_os = "macos",
+            target_os = "android",
+            target_env = "musl",
+            target_pointer_width = "64"
+        ),
+        not(target_os = "freebsd")
+    ))]
+    pub fn files_free(&self) -> libc::uint64_t {
+        self.0.f_ffree
+    }
+
+    /// Free file nodes in filesystem
+    #[cfg(all(
+        target_pointer_width = "32",
+        not(any(
+            target_os = "ios",
+            target_os = "macos",
+            target_os = "android",
+            target_os = "freebsd",
+            target_env = "musl"
+        ))
+    ))]
+    pub fn files_free(&self) -> libc::uint32_t {
+        self.0.f_ffree
+    }
+
+    /// Free file nodes in filesystem
+    #[cfg(target_os = "freebsd")]
+    pub fn files_free(&self) -> i64 {
+        self.0.f_ffree
+    }
+
+    /// Filesystem ID
+    pub fn filesystem_id(&self) -> fsid_t {
+        self.0.f_fsid
+    }
+}
+
+pub fn statfs<P: ?Sized + NixPath>(path: &P) -> Result<Statfs> {
+    unsafe {
+        let mut stat: Statfs = mem::uninitialized();
+        let res = path.with_nix_path(|path| libc::statfs(path.as_ptr(), &mut stat.0))?;
+        Errno::result(res).map(|_| stat)
     }
 }
 
-pub fn fstatfs<T: AsRawFd>(fd: &T, stat: &mut libc::statfs) -> Result<()> {
+pub fn fstatfs<T: AsRawFd>(fd: &T) -> Result<Statfs> {
     unsafe {
-        Errno::clear();
-        Errno::result(libc::fstatfs(fd.as_raw_fd(), stat)).map(drop)
+        let mut stat: Statfs = mem::uninitialized();
+        Errno::result(libc::fstatfs(fd.as_raw_fd(), &mut stat.0)).map(|_| stat)
+    }
+}
+
+#[cfg(test)]
+#[cfg(not(target_os = "ios"))]
+mod test {
+    use std::fs::File;
+    use sys::statfs::*;
+
+    #[test]
+    fn statfs_call() {
+        statfs("/tmp".as_bytes()).unwrap();
```

Comment:
These tests don't provide much coverage.  Could you check the output somehow?  One way would be to compare the output of `statfs` with `statvfs`.